sdk/helper/certutil/helpers.go

```diff
@@ -863,16 +858,25 @@ func createCertificate(data *CreationBundle, randReader io.Reader, privateKeyGen
 }
 
 func selectSignatureAlgorithmForECDSA(pub crypto.PublicKey, signatureBits int) x509.SignatureAlgorithm {
-	// If signature bits are configured, prefer them to the default choice.
-	switch signatureBits {
-	case 256:
-		return x509.ECDSAWithSHA256
-	case 384:
-		return x509.ECDSAWithSHA384
-	case 512:
-		return x509.ECDSAWithSHA512
-	}
-
+	// Previously we preferred the user-specified signature bits for ECDSA
+	// keys. However, this could result in using a longer hash function than
+	// the underlying NIST P-curve will encode (e.g., a SHA-512 hash with a
+	// P-256 key). This isn't ideal: the hash is implicitly truncated
+	// (effectively turning it into SHA-512/256) and we then need to rely
+	// on the prefix security of the hash. Since both NIST and Mozilla guidance
+	// suggest instead using the correct hash function, we should prefer that
+	// over the operator-specified signatureBits.
+	//
+	// Lastly, note that pub above needs to be the _signer's_ public key;
+	// the issue with DefaultOrValueHashBits is that it is called at role
+	// configuration time, which might _precede_ issuer generation. Thus
+	// it only has access to the desired key type and not the actual issuer.
+	// The reference from that function is reproduced below:
+	//
+	// > To comply with BSI recommendations Section 4.2 and Mozilla root
```

Comment:
Confused looking at this list of curves below: we explicitly set ECDSAWithSHA256 for P224 and 256 (but not P192 which is set by default)?

I think it would be cleaner/safer if we added P192 to the first case, then either error if we don't recognize the alrogirthm?

Reply:
We actually restrict it to at least P-224:

https://github.com/hashicorp/vault/blob/74248e1b20a6f48bfc94f7d72551fc5cebdbd3ea/sdk/helper/certutil/helpers.go#L690-L694 

And Go lacks support for P-192: https://github.com/golang/go/blob/master/src/crypto/elliptic/elliptic.go#L442-L447 and https://pkg.go.dev/crypto/elliptic

So there's no way to do this comparison.

What is the hash function for P-192 anyways? [This](https://csrc.nist.gov/csrc/media/projects/cryptographic-standards-and-guidelines/documents/examples/ecdsa_prime.pdf) makes me think it is the insecure SHA-1 which isn't... good? :-)